records/chains.go

```diff
@@ -0,0 +1,80 @@
+package records
+
+import (
+	"github.com/mesosphere/mesos-dns/records/labels"
+)
+
+// chain is a generation func that consumes record-like strings and does
+// something with them
+type chain func(...string)
+
+const (
+	protocolNone = "" // for readability
+	domainNone   = "" // for readability
+)
+
+// withProtocol appends `._{protocol}.{framework}` to records. if protocol is "" then
+// the protocols "tcp" and "udp" are assumed.
+func withProtocol(protocol, framework string, spec labels.Func, gen chain) chain {
+	return func(records ...string) {
+		protocol = spec(protocol)
+		if protocol != protocolNone {
+			for i := range records {
+				records[i] += "._" + protocol + "." + framework
+			}
+		} else {
+			records = append(records, records...)
+			for i, j := 0, len(records)/2; j < len(records); {
+				records[i] += "._tcp." + framework
+				records[j] += "._udp." + framework
+				i++
+				j++
+			}
+		}
+		gen(records...)
+	}
+}
+
+// withSubdomains appends `.{subdomain}` (for each subdomain spec'd) to records.
+// the empty subdomain "" indicates to generate records w/o a subdomain fragment.
+func withSubdomains(subdomains []string, gen chain) chain {
+	if len(subdomains) == 0 {
+		return gen
+	}
+	return func(records ...string) {
+		var (
+			recordLen = len(records)
+			tmp       = make([]string, recordLen*len(subdomains))
+			offset    = 0
+		)
+		for s := range subdomains {
```

Comment:
it probably doesn't matter for this case. in general, i'm trying to be more conservative with `range` statements over slices that use two iteration vars (index, value) because i've recently been bitten several times by the fact that `value` ends up being a copy of what's in the slice. so while it probably doesn't matter here, i don't see any reason to change the code - it's still very readable